Implement LTI config by URL (#1054)

lsloan opened the pull request:
Resolves #1054.

I'm working on wiki docs for this new feature.

## Changes based on code review
* [x] Move the LTI config template out of `config/env_sample.json` and into a new file, `assets/config/lti_config_template.json`.  See: https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1105#discussion_r502669168
* [x] Change `"target_link_uri": "{base_url}{url_login}"` in template from `url_login` to `url_launch`.  See: https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1105#discussion_r505511939


## Testing
More testing details in the issue, #1054. 
1. Go to Canvas' LTI Developer Key creation page:
    * https://umich.instructure.com/accounts/1/developer_keys#lti_key_modal_opened
1. In the "Method" menu, select the "Enter URL" item.
1. Copy the following URL and paste it into the "JSON URL" field:
    * https://nnnnnnnnnnnn.ngrok.io/lti/config/
    * Replace "nnnnnnnnnnnn" with the ID number of your ngrok server.
1. Copy the following URL and paste it into the "Redirect URIs" field:
    * https://nnnnnnnnnnnn.ngrok.io/lti/launch/
    * Replace "nnnnnnnnnnnn" with the ID number of your ngrok server.
1. (Optional, recommended) Enter meaningful values into the "Key Name", "Owner Email", and "Notes" fields.
1. Click the "Save" button.

jonespm commented on config/env_sample.json:
```diff
@@ -117,6 +117,53 @@
             }
         ]
     },
+    "LTI_CONFIG_TEMPLATE": {
```
It doesn't look like this is something anyone would have to ever modify and everything is replaced in the code.

If it's just going to be a static json template we should probably just have it out in it's own file, part of the app separate from this json and config.

lsloan replied:
I wholeheartedly agree!

I [asked on Slack for opinions](https://app.slack.com/client/T69CPPGNP/GNSJ3RMHN/thread/GNSJ3RMHN-1602174463.002200) about where the JSON template should go.

I wanted to support HJSON for the template file to allow comments in it.  But when I thought about the places in the project where the file could go, the `secrets` directory came to mind.  Since `env.json` already existed there, I thought it might as well go there.  It doesn't support HJSON (see: #1056), unfortunately, but it seems like a good place for it.

I also thought that there's a chance the format of the LTI config JSON could change in the future.  Either because of the way LTI is implemented or some of the Canvas-related values change.  So putting the LTI config template in the main app config makes it easier for others to update it as necessary.

Suggest a place for the template file to go and I'll put it there.

jonespm replied:
I see, I haven't been looking at slack (outside of prod support) this week to focus on data studio. I see what you were talking about though. To me it feel like it should go in the dashboard folder as it's part of the app. Assets is where the current static file are stored (when collect static is run).

I'd probably put it in like `assets/config` or `assets/json` ? 

Then you should be able to just read it with `static("config/lti_config_template.json") `

pushyamig replied:
Yeah, I agree that the JSON template right now only covers the properties that we us in MyLA and we might want to add based on the needed. For Example `"default": "disabled",` this was added later as we did not want tool to be installed by default

pushyamig replied:
Having this as a configuration option should be a way to go `config/lti_config_template.json` so that in case of changes we can just change the configuration. But there might be certain options we might associate with the code

pushyamig replied:
Their might be certain LTI configuration options we might not have used might use in future having it configuration option is good and having it as separate file(`lti_config_template.json`) is even better so that it doesn't clutter the env.json as that file purpose is purely for properties that differ by env

lsloan replied:
@pushyamig, I'm confused.

Which do you mean?

* [ ] Move LTI config template to `config/lti_config_template.json`
* [ ] Keep LTI config in `config/env_sample.json` (and therefore, the deployed `/secrets/env.json`)
* [ ] something else

lsloan replied:
Thanks for clarifying!  I'll move it to `config/lti_config_template.json`.

lsloan replied:
As I move this to another file, I had initially thought I would take the opportunity to add HJSON support to this.

However, as I think about it more, I have an opportunity to simplify this code.  Supporting HJSON for commenting and formatting probably won't be important for this file.  The comments would need to be parsed out before it could be sent to the requesting LMS anyway.  That would just add another step to the current process of replacing all the parameters in the string values and serializing everything back to JSON.

Instead, my thinking is to just use the new JSON file contents as one big string, apply formatting to it with the defined parameters, and return that to the LMS.  It wouldn't need the `recursive_format()` function at all.  That's a little disappointing, but it's probably worthwhile to reduce the complexity of the code.

lsloan replied:
> Then you should be able to just read it with `static("config/lti_config_template.json") `

There were a few implementations of `static()` in Django, but none of them gave me the path to the file or its contents, as I was expecting.

In the documentation for Django's `staticfiles` app (https://docs.djangoproject.com/en/3.1/ref/contrib/staticfiles/#finders-module), I found that `finders.find()` would give me the full path to the file, so I use that like this:

```python
open(finders.find('config/lti_config_template.json')).read()
```

lsloan commented:
FYI: My biggest regret with this feature is that I couldn't make the string formatting work in a subclass of [`JSONEncoder`](https://docs.python.org/3.8/library/json.html?highlight=jsonencoder#json.JSONEncoder).  It's fairly easy to add support for encoding classes that are currently unsupported.  Trying to change how it encodes strings, a fundamental part of JSON, was elusive.

I decided to give in and go with this pre-encoding step instead.

I'd still welcome any ideas of how to make a JSONEncoder subclass solution work, though!

Also, it just occurred to me, maybe the parameter values (other than timestamp) should be filled into the template when it's being *decoded* from the config file instead.  Maybe `JSONDecoder` stands a better chance of handling this easily, although not as flexibly as I'd like (loss of timestamp, etc.).

pushyamig commented on dashboard/lti_new.py:
```diff
@@ -75,6 +79,48 @@ def get_jwks(request):
+
+def generate_config_json(request: HttpRequest) -> Union[
+    JsonResponse, LTIError]:
+    config = get_tool_conf()
```
What is the need for doing this? This is something PyLTI1.3 library after grabbing the LTI_CONFIG values try to do the LTI login/launch. when we configure the `/lti/config/` Myla simply return the JSON.

lsloan replied:
Because the LTI should be enabled and configured properly before an LMS requests the config and tries to use it.

That seems important to me.

lsloan replied:
@pushyamig (or @jonespm), can you confirm whether checking MyLA's LTI config before generating the config for the LMS is OK, or does it cause a problem?

pushyamig replied:
I tested it locally and it did create configuration and I pointed those config in my local instance and it worked as expected. I did that yesterday and not with updated changes. Should work even with recent changes as it is not something drastically different

pushyamig replied:
@lsloan I don't see any harm done, it just instantiate the tool config. I personally would have not done it. Since this is purely communication between Myla and Canvas to get the LTI keys. Even let's say if we did not put any LTI_CONFIGS then it will throw errors. I don't know if Canvas notifies with the response we are sending or defaults to its own error message. I am fine either way

lsloan replied:
To test what happens when MyLA's LTI configuration is invalid, I altered the code a little to simulate that.

## Results

### Canvas

Canvas showed the message

> Request failed with status code 422

MyLA's response actually has a status code of 500.  I believe 422 is the response from Canvas' internal processing to its UI.  (422 is "Unprocessable Entity", probably the error given when Canvas didn't find the expected content in the LTI config JSON.)

That message is good enough to tell the user something was wrong with MyLA.

### MyLA

MyLA logged an error message.  I'm making some changes to that log message now, to be a little more specific.

> [2020-10-16 14:58:16 -0400] [ERROR] [lti_new.py:72] Invalid LTI configuration: "0"

That message should help MyLA administrators that something is wrong.

The content of the response also includes the error.

```json
{"lti_error": "Invalid LTI configuration: \"0\""}
```

## Conclusion

It's helpful to check the LTI configuration, even if it's not being used in that block of code.

pushyamig replied:
Thanks for the detailed text. I guess Canvas will show the error response based on it's interpretation and it is fine that way. +1 with log message

pushyamig commented on config/env_sample.json:
```diff
@@ -117,6 +117,53 @@
         ]
     },
+    "LTI_CONFIG_TEMPLATE": {
+        "title": "MyLA autoconfig-{timestamp}",
```
If we don't specify the `KeyName` in the LTI dev key, will this value here replaced their?

pushyamig replied:
I got confused with KeyName and title.

pushyamig commented on dashboard/lti_new.py:
```diff
@@ -75,6 +79,48 @@ def get_jwks(request):
+            [request.scheme, request.get_host()] + 3 * ['']),
+        'url_login': reverse('login'),
+        'url_launch': reverse('launch'),
+        'url_get_jwks': reverse('get_jwks'),
```
I like the way you were autofilling this rather me changing it manual in config

lsloan replied:
Thanks.

lsloan commented:
@jonespm and @pushyamig, please review this again and tell me if you find problems or give it an approval so I can merge it.

I've fixed a mistake in the `target_link_uri` property of the template and moved the template to a new file, which also simplified the logic.

pushyamig commented on dashboard/lti_new.py:
```diff
@@ -64,17 +69,40 @@ def check_if_success_getting_tool_config(tool_config):
+
+    return HttpResponse(
+        open(finders.find(
+            'config/lti_config_template.json')).read() % parameters,
```
I though we want to upload the `lti_config_template.json` as secret  as in `env.json`. so that we can change it on fly if we have to. Because all the file that are place in `config` are copied to a local `mylasecrets` and in docker-compose `${HOME}/mylasecrets:/secrets`. 
@jonespm Thoughts???

lsloan replied:
@pushyamig, if that's what you meant, I didn't understand your comments.

In my comment at (https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1105#discussion_r505579885), I asked for clarification.  You chose the "Move LTI config template to assets/config/lti_config_template.json" option and gave the comment a "thumbs up" reaction.

I also repeated the option in the comment at (https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1105#discussion_r505583005) so there would not be any misunderstanding.

Putting the template into a secret is overkill.  It doesn't contain anything that would be considered "secret" and it's unlikely to be changed.

pushyamig replied:
@lsloan Sorry for confusing comment. But Please keep the file as in the config directory and it's fine. I do agree we don't change the template file that often and if we do have to change something in emergency we can do it in canvas directly. This template is only used during initial creating time of LTI Key.

pushyamig commented on dashboard/lti_new.py:
```diff
@@ -64,17 +69,40 @@ def check_if_success_getting_tool_config(tool_config):
+    return HttpResponse(
+        open(finders.find(
+            'config/lti_config_template.json')).read() % parameters,
+        content_type='application/json')
```
`open(finders.find(` this statement seems to directly opening the file and making an assumption the we will find that file there and exception handling/logging are not in place? 
Just before rendering as HTTPResponse we need to check if file exists and params are rightly applied and log accordingly stating the issue

lsloan replied:
Yes, it does use the file without checking for errors.

If you think it's necessary to check for errors reading the file, I can add that.  I didn't think it was absolutely necessary because the file is being read from the project folders, as suggested by @jonespm.  That is, we have more control over the file than if it were in `/secrets`.

lsloan commented:
I appreciate the approval, @pushyamig.

I've been thinking about this issue this morning.

Although I don't think putting the template inside the env.json or a file in /secrets is the right thing to do, I think it wouldn't hurt if we provide some kind of "escape hatch" for admins setting this up, in case they need to change the template.  (To address a LTI spec change, etc.)

So, I think I'll add an option to env.json that will normally be null, but if it's not and it contains a pathname to an valid alternate template file, it will be used instead.  That will require a little more error handling around the use of the file, but I think it will be worth it.
